File: First Year/Second Semester/Graph Algorithms/Assignment1CPP/GraphGenerator/GraphRandomGenerator.cpp
//
// Created by Emanuel on 28.03.2021.
//

#include "GraphRandomGenerator.h"

#include <vector>
#include <algorithm>

DirectedGraph GraphRandomGenerator::generateGraph(const int &vertices, const int &edges) const {
    if (edges > vertices * vertices)
        throw std::string("Invalid number of edges (too many).\n");

    std::vector<std::tuple<int, int, int>> all_edges;
    for (size_t x = 0; x < vertices; ++x)
        for (size_t y = 0; y < vertices; ++y) {
            int cost = rand() % 100;
            all_edges.push_back({x, y, cost});
        }
    int shuffle_cnt = rand() % 10;
    for (size_t i = 0; i < shuffle_cnt; ++i)
        std::random_shuffle(all_edges.begin(), all_edges.end());

        DirectedGraph graph {vertices};
        for (size_t i = 0; i < edges; ++i)
            graph.addEdge(std::get<0>(all_edges[i]), std::get<1>(all_edges[i]), std::get<2>(all_edges[i]));
        return graph;
}

File: First Year/Second Semester/Graph Algorithms/Assignment2/GraphGenerator/GraphRandomGenerator.py
from random import randint, shuffle
from Domain.DirectedGraph import DirectedGraph


class GraphRandomGenerator:
    @staticmethod
    def generateGraph(vertices, edges):
        """
            Generates a random graph
            :param vertices: number of vertices for the generated graph
            :param edges: number of edges for the generated graph
            :return: the generated graph
            :raises: ValueError if the graph cannot be generated
        """
        if edges > vertices * vertices:
            raise ValueError("Invalid number of edges (too many).\n")

        all_edges = []
        for x in range(vertices):
            for y in range(vertices):
                cost = randint(0, 100)
                all_edges.append((x, y, cost))
        shuffle(all_edges)

        graph = DirectedGraph(vertices)
        for i in range(edges):
            graph.addEdge(all_edges[i][0], all_edges[i][1], all_edges[i][2])
        return graph
